### ISSUE TITLE
Upgrade from 2.0 -> 2.1 Issues

### ISSUE THREAD

**Issue by jessecureton** [papermerge/papermerge-core#111]
I've been trying to get my papermerge 2.0.1 installation updated to the new 2.1 release, and having lots of issues. The root of the issue seems to be that the `core_user.id` column in my old database is an integer, when some migrations are expecting a `uuid`. This means that 9 migrations fail to apply when first launching, and from there everything just seems to cascade and I can't sign in due to more errors with the core_user table when attempting a login.

Is this an upgrade that's expected to work? Right now things seem pretty borked, but I'm happy to pull a DB backup and help gather more data if this is something that's expected to work

**Comment by ciur** [comment]
Thank you for opening this ticket.
In short migration from 2.0 to 2.1 is more complex then just DB changes. Just DB part will not work even if you succeed in fixing db migrations; why ? because 2.1 introduces new model for document version which was there before :).

A picture will help understanding why DB changes won't work:

![schema-changes](https://user-images.githubusercontent.com/24827601/210509157-e6deb4bb-d208-4bcb-92fc-fc39167cb38f.png)

As you see from the picture, the "Document Version" model was introduced in 2.1 (before that, document versions were "stored" only on file-system and there is literally no information about document version(s) in DB)

Another important change is actually what information is stored on the file-system level i.e. what and how OCRed data is stored
on file system. "OCR data" is much more then just plain text characters - extracted "OCR data" also stores coordinates of each extracted character (this info has a fancy name - [hocr](https://en.wikipedia.org/wiki/HOCR) format).

Long story short - at this moment you don't need to do anything.
Please give me couple of weeks and I will document and provide necessary tools on how to migrate you documents from 2.0 to 2.1.

**Comment by jessecureton** [comment]
Hi @ciur! Thanks again for all your work on this. I saw the newest release and wanted to follow up on this and see if there has been any progress on a migration path from 2.0->2.1?

**Comment by duracell** [comment]
Hey @ciur, any news about the migration to 2.1? I'm currently using the old linuxserver image (2.0x), but would love to switch over and upgrade to a new version. Should it work?

**Comment by ReneHezser** [comment]
Hi @ciur 
I am looking for upgrade instructions from 2.0 to the latest 3.2 version and was wondering if there is an updated upgrade instruction available. The current installation is running in a container and is installed with Docker Compose.

**Comment by duracell** [comment]
It really seems like a mess :'/
No updates for old versions, no migration to new version and totally mixed features among different versions, without any explanation if this features will be return or not. The great sharing feature was the reason why I choose papermerge over Paperless-ng(x), but it seems the new version just getting the features of paperless and drops the unique features. :(
I really greatful and thank ciur for the fantastic work of this software, but this great work makes me even more sad that I'm using an old version and can't update =/

**Comment by ciur** [comment]
> It really seems like a mess :'/ Breaking update from 2 → 2.1 and another breaking update from 2.1 → 3.0…

Yes, I agree, it is confusing and painful to have those breaking incompatibilities. When I've started this project I had no idea how to write a DMS software and as result I made terrible mistakes in design of the application. The outcome was very painful maintenance of the code base. Also adding new features was hard because of the base structure on which the entire project was build.

Because I have big plans for the project and because the project is still relatively new I've decided to break backward compatibility for the sake of fixing design errors in early stage of the project.

The good part is that since 3.0 there won't be any breaking changes in terms of how files are stored on the disk and database schema: this means that once in 3.0 you will be able to update to 3.1, 3.2, 3.3,... without any problem.

> No updates for old versions, no migration to new version and totally mixed features among different versions, without any explanation if these features will be return or not.

All great features present in 2.0 (like sharing files between users, automates) will be back. The "sharing of the files between users" will be back this year (2024).

> but it seems the new version just getting the features of paperless and drops the unique features

In short term yes - because of large code base refactoring inevitably some of the non essential features need to be "temporary removed". But in long term, features from 2.0 will be back with much more.

> I'm using an old version and can't update.

I am currently in deep dive with 3.3; when 3.3. released, I will spend time documenting migration from 2.0 and 2.1 directly into 3.x.

**Comment by ReneHezser** [comment]
Thank you very much for you effort on this great project.

> I am currently in deep dive mode with 3.3; with 3.3. released, I will dedicate time for documenting migration from 2.0 and 2.1 directly into 3.x.

That is the important information. At least I can and will wait some more time to upgrade later. Thanks.

**Comment by duracell** [comment]
Thanks for the fast and informative reply and the promise of a stable future. I think that is really important for a DMS.

> I am currently in deep dive mode with 3.3; with 3.3. released, I will dedicate time for documenting migration from 2.0 and 2.1 directly into 3.x.

That sounds great, and then I will wait instead of trying to migrate everything to another software, as there is currently no real pressure except the missing updates.

**Comment by duracell** [comment]
@ciur any news about upgrade/migrate an installation to a new version?

**Comment by duracell** [comment]
So, is the unpinning a message?
I just would like to know if it's useful to wait or if I just should manual migrate to new version/software.
A short info would be helpful, even if you just say you don't have the resources to do this documentation.

**Comment by ciur** [comment]
Migrate to the newest version  which is 3.3.
I don't have resources to do this documentation.

**Comment by ciur** [comment]
@duracell Are you on 2.0?

Is the migration problem still relevant ?

The plan now is to provide migration scripts from 2.0 to 3.4. So I will update this ticket title as "migrate from 2.0 to 3.4".
Will be ok for you to migrate directly 3.4 ?
